File: backend/routers/audit.py
"""
Router de Auditoria - xFinance

🔒 ADMIN ONLY: Todas as rotas requerem papel de administrador.

Endpoints:
- GET  /api/audit/{id_princ}  - Histórico de um registro
- GET  /api/audit/stats       - Estatísticas do log
- POST /api/audit/cleanup     - Limpar registros expirados
"""

import logging
from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel

from dependencies import CurrentUser, require_admin
from services.audit import get_history, cleanup_expired, get_stats

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/audit", tags=["Audit"])


# =============================================================================
# SCHEMAS
# =============================================================================

class AuditEntry(BaseModel):
    """Entrada do log de auditoria."""
    id_log: int
    id_user: int
    user_email: str
    operacao: str
    campo: Optional[str] = None
    valor_anterior: Optional[str] = None
    valor_novo: Optional[str] = None
    dt_operacao: str


class AuditHistoryResponse(BaseModel):
    """Response do histórico de auditoria."""
    id_princ: int
    total: int
    entries: List[AuditEntry]


class AuditStatsResponse(BaseModel):
    """Response das estatísticas de auditoria."""
    total_registros: int
    registro_mais_antigo: Optional[str] = None
    registros_expirados: int


class CleanupResponse(BaseModel):
    """Response da limpeza de registros."""
    success: bool
    message: str
    deleted: int


# =============================================================================
# GET /api/audit/stats - Estatísticas (deve vir antes de /{id_princ})
# =============================================================================

@router.get("/stats", response_model=AuditStatsResponse)
async def audit_stats(
    current_user: CurrentUser = Depends(require_admin),
):
    """
    Retorna estatísticas do log de auditoria.
    
    🔒 ADMIN ONLY
    """
    logger.info("GET /audit/stats | user=%s", current_user.email)
    
    stats = get_stats()
    return AuditStatsResponse(**stats)


# =============================================================================
# POST /api/audit/cleanup - Limpar expirados
# =============================================================================

@router.post("/cleanup", response_model=CleanupResponse)
async def audit_cleanup(
    current_user: CurrentUser = Depends(require_admin),
):
    """
    Remove registros de auditoria expirados (mais de 14 meses).
    
    🔒 ADMIN ONLY
    """
    logger.info("POST /audit/cleanup | user=%s", current_user.email)
    
    try:
        deleted = cleanup_expired()
        
        if deleted > 0:
            return CleanupResponse(
                success=True,
                message=f"{deleted} registro(s) antigo(s) removido(s)",
                deleted=deleted,
            )
        else:
            return CleanupResponse(
                success=True,
                message="Nenhum registro expirado para remover",
                deleted=0,
            )
            
    except Exception as e:
        logger.error("Erro ao limpar auditoria: %s", e)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Erro ao limpar registros de auditoria"
        )


# =============================================================================
# GET /api/audit/{id_princ} - Histórico de um registro
# =============================================================================

@router.get("/{id_princ}", response_model=AuditHistoryResponse)
async def audit_history(
    id_princ: int,
    limit: int = Query(100, ge=1, le=500),
    current_user: CurrentUser = Depends(require_admin),
):
    """
    Retorna histórico de operações de um registro específico.
    
    🔒 ADMIN ONLY
    
    Args:
        id_princ: ID do registro (tabela princ)
        limit: Limite de registros (padrão 100, máx 500)
    """
    logger.info(
        "GET /audit/%d | user=%s | limit=%d",
        id_princ, current_user.email, limit
    )
    
    try:
        entries = get_history(id_princ, limit)
        
        return AuditHistoryResponse(
            id_princ=id_princ,
            total=len(entries),
            entries=[AuditEntry(**entry) for entry in entries],
        )
        
    except Exception as e:
        logger.error("Erro ao buscar histórico: %s", e)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Erro ao buscar histórico de auditoria"
        )

File: client/src/services/api/audit.ts
/**
 * API de Auditoria - xFinance
 * 
 * Serviço para consulta do histórico de operações (audit log).
 * Todas as rotas requerem papel: admin
 */

// =============================================================================
// TIPOS
// =============================================================================

export interface AuditEntry {
  id_log: number;
  id_user: number;
  user_email: string;
  operacao: "CREATE" | "UPDATE" | "DELETE" | "ENCAMINHAR";
  campo: string | null;
  valor_anterior: string | null;
  valor_novo: string | null;
  dt_operacao: string;
}

export interface AuditHistoryResponse {
  id_princ: number;
  total: number;
  entries: AuditEntry[];
}

export interface AuditStatsResponse {
  total_registros: number;
  registro_mais_antigo: string | null;
  registros_expirados: number;
}

export interface CleanupResponse {
  success: boolean;
  message: string;
  deleted: number;
}

// =============================================================================
// API BASE URL
// =============================================================================

const API_BASE = "";

// =============================================================================
// FUNÇÕES
// =============================================================================

/**
 * Busca histórico de operações de um registro específico.
 * Requer papel: admin
 */
export async function fetchAuditHistory(
  idPrinc: number,
  limit: number = 100
): Promise<AuditHistoryResponse> {
  const response = await fetch(
    `${API_BASE}/api/audit/${idPrinc}?limit=${limit}`,
    { credentials: "include" }
  );

  if (response.status === 401) {
    throw new Error("Não autenticado");
  }

  if (response.status === 403) {
    throw new Error("Acesso restrito a administradores");
  }

  if (!response.ok) {
    const error = await response.json().catch(() => ({}));
    throw new Error(error.detail || "Erro ao buscar histórico");
  }

  return response.json();
}

/**
 * Retorna estatísticas do log de auditoria.
 * Requer papel: admin
 */
export async function fetchAuditStats(): Promise<AuditStatsResponse> {
  const response = await fetch(`${API_BASE}/api/audit/stats`, {
    credentials: "include",
  });

  if (response.status === 401) {
    throw new Error("Não autenticado");
  }

  if (response.status === 403) {
    throw new Error("Acesso restrito a administradores");
  }

  if (!response.ok) {
    const error = await response.json().catch(() => ({}));
    throw new Error(error.detail || "Erro ao buscar estatísticas");
  }

  return response.json();
}

/**
 * Remove registros de auditoria expirados (mais de 14 meses).
 * Requer papel: admin
 */
export async function cleanupAudit(): Promise<CleanupResponse> {
  const response = await fetch(`${API_BASE}/api/audit/cleanup`, {
    method: "POST",
    credentials: "include",
  });

  if (response.status === 401) {
    throw new Error("Não autenticado");
  }

  if (response.status === 403) {
    throw new Error("Acesso restrito a administradores");
  }

  if (!response.ok) {
    const error = await response.json().catch(() => ({}));
    throw new Error(error.detail || "Erro ao limpar registros");
  }

  return response.json();
}
